### PR TITLE
[jsonrpc] support Player.AddSubtitle API

### PR DESCRIPTION
**Pull request by pespin**
This allows adding and applying a new subtitle URI to the player, be it
a URL, a local file, etc.

## Description
This allows adding and applying a new subtitle URI to the player, be it
a URL, a local file, etc.

## Motivation and Context
Be able to set the subtitle from a URL, see #18443 

## How Has This Been Tested?
I tested building kodi master with the patch applied, then using json-rpc to play a video from an http server using castkodi firefox addon. Then I used kodi's web interface API browser to set a .srt file successfully (/#lab/api-browser/Player.AddSubtitle), both from a local file (/tmp/bla.srt) and from an HTTP address.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ X ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ X ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by pespin** [comment]
Can somebody give me a hint on what's needed to have the "Mergeable" thing working?

**Comment by garbear** [comment]
> Can somebody give me a hint on what's needed to have the "Mergeable" thing working?

You can click "details" and see the results: https://github.com/xbmc/xbmc/pull/18445/checks?check_run_id=1139208101. In this case there was some github fluff missing.

I added some labels and a milestone and it's green now.

**Comment by DaveTBlake** [comment]
When making a non-breaking change to the JSON API schema such as adding parameters or new methods, the minor number needs to be bumped (and patch set to 0) in version.txt https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/version.txt . The schema documents this, see
https://github.com/xbmc/xbmc/blob/6fdce600bb850ecd379d3ff63c98fd600279e937/xbmc/interfaces/json-rpc/schema/methods.json#L33

**Comment by CastagnaIT** [comment]
can you give an full example of usage?

**Comment by pespin** [comment]
@CastagnaIT I think I already provided a full example of usage in the PR description as well as in https://github.com/xbmc/xbmc/issues/18443

See also my branch implementing use for this new API in castkodi firefox addon:
https://github.com/pespin/castkodi/commit/6189e353db026a24b3a92849ac3648a0f3450512

**Comment by CastagnaIT** [comment]
i see the json rpc call 

>         return this.kodi.send("Player.AddSubtitle", {
>             playerid: 1,
>             subtitle:  subtitle,
>         });

but i do not understand well what data is passed to "subtitle" variable
i suppose the file path e.g. "C:\sub\sub.srt"

if so, the name of the json rpc argument "subtitle" can be confusing
it is better rename with something like "filepath"?

**Comment by pespin** [comment]
@CastagnaIT the parameter can be whatever string kodi can reproduce afaict. I successfully tested passing a URL (http://192.168.1.100/bla.srt) or a local file (/tmp/bla.srt).

That means I can basically have some web directory with a media file and an external srt and simply pass that to to kodi to stream and load the subtitles without having to copy over kodi's host manually the srt file.

**Comment by pespin** [comment]
So is there any concern about this PR, or can it be merged?

File: xbmc/interfaces/json-rpc/schema/methods.json [hunk 1]
```diff
@@ -457,6 +457,17 @@
...
     "returns": "string"
   },
+  "Player.AddSubtitle": {
+    "type": "method",
+    "description": "Add subtitle to the player",
+    "transport": "Response",
+    "permission": "ControlPlayback",
+    "params": [
+      { "name": "playerid", "$ref": "Player.Id", "required": true },
+      { "name": "subtitle", "type": "string", "required": true }
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/schema/methods.json:467]: Please add a description. Is this a path to a local / external file?

File: xbmc/interfaces/json-rpc/PlayerOperations.cpp [hunk 2]
```diff
@@ -1104,6 +1104,29 @@ JSONRPC_STATUS CPlayerOperations::SetAudioStream(const std::string &method, ITra
   return ACK;
 }
 
+JSONRPC_STATUS CPlayerOperations::AddSubtitle(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  switch (GetPlayer(parameterObject["playerid"]))
+  {
+    case Video:
+        break;
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PlayerOperations.cpp:1112]: This looks odd. Either just use `if (GetPlayer(parameterObject["playerid"]) != Video) return FailedToExecute;` or move the actual implementation into the `switch-case` as done in `SetAudioStream()` et al.

**Review by Montellese — CHANGES_REQUESTED** [review]
Looks good apart from the minors.

**Comment by DaveTBlake** [comment]
The version update has been lost in the recent rebase @pespin 

When making a non-breaking change to the JSON API schema such as adding parameters or new methods, the **minor number** needs to be bumped (and patch set to 0) in version.txt https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/version.txt . The schema documents this, see
https://github.com/xbmc/xbmc/blob/6fdce600bb850ecd379d3ff63c98fd600279e937/xbmc/interfaces/json-rpc/schema/methods.json#L33

**Comment by DaveTBlake** [comment]
jenkins build this please

**Review by DaveTBlake — APPROVED** [review]

**Comment by pespin** [comment]
Hi,
I cannot find where/how jenkins failed. Can it be merged? Do I need to do something?

**Comment by jakkerson** [comment]
Since this PR merge (25th) built in subs on files no longer display.

On the 23rd Nightly subs built into files display fine, 25th and after no subs display.

I can get around to an example with log at a later date, but if someone else is able to test before I can that would be great.

**Comment by Tolriq** [comment]
@jakkerson This PR can't be the source of that problem, it's self contained and have 0 impact if nothing calls the new JSON endpoint that no one probably does for the moment.

**Comment by jakkerson** [comment]
Sorry you are correct, it was due to Forced subtitle tracks not displaying any subs with English audio. Ignore my comment.

**Pull request merged**